packages/zoe/src/contractFacet/contractFacet.js

```diff
@@ -0,0 +1,196 @@
+        const previousAmounts = flattenAllocations(previousAllocations);
+
+        const newAllocations = seatStagings.map(seatStaging =>
+          seatStaging.getStagedAllocation(),
+        );
+        const newAmounts = flattenAllocations(newAllocations);
+
+        assert(
+          areRightsConserved(getAmountMath, previousAmounts, newAmounts),
+          details`Rights are not conserved in the proposed reallocation`,
+        );
+
+        // Commit the staged allocations and inform Zoe of the
+        // newAllocation.
+        seatStagings.forEach(seatStaging =>
+          seatToSeatAdmin.get(seatStaging.getSeat()).commit(seatStaging),
+        );
+      },
+      saveIssuer: (issuerP, keyword) => {
+        return E(zoeInstanceAdmin)
+          .saveIssuer(issuerP, keyword)
+          .then(() => {
+            return issuerTable
+              .getPromiseForIssuerRecord(issuerP)
+              .then(issuerRecord => {
+                assertKeywordName(keyword);
+                assert(
+                  !getKeywords(instanceRecord.issuerKeywordRecord).includes(
+                    keyword,
+                  ),
+                  details`keyword ${keyword} must be unique`,
+                );
+                instanceRecord.issuerKeywordRecord = {
+                  ...instanceRecord.issuerKeywordRecord,
+                  [keyword]: issuerRecord.issuer,
+                };
+                instanceRecord.brandKeywordRecord = {
+                  ...instanceRecord.brandKeywordRecord,
+                  [keyword]: issuerRecord.brand,
+                };
+
+                return issuerRecord;
+              });
+          });
+      },
+      makeInvitation: (offerHandler, description, customProperties = {}) => {
+        assert.typeof(
+          description,
+          'string',
+          details`invitations must have a description string: ${description}`,
+        );
+        /** @type {InvitationHandle} */
+        const invitationHandle = {};
+        harden(invitationHandle);
+        invitationHandleToHandler.init(invitationHandle, offerHandler);
+        /** @type {Promise<Payment<'ZoeInvitation'>>} */
+        const invitationP = E(zoeInstanceAdmin).makeInvitation(
+          invitationHandle,
+          description,
+          customProperties,
+        );
+        return invitationP;
+      },
+      // Shutdown the entire vat and give payouts
+      shutdown: () => E(zoeInstanceAdmin).shutdown(),
+
+      // The methods below are pure and have no side-effects //
+      getZoeService: () => zoeService,
+      getInvitationIssuer: () => invitationIssuer,
+      getInstanceRecord: () => harden({ ...instanceRecord }),
+      getBrandForIssuer: issuer =>
+        issuerTable.getIssuerRecordByIssuer(issuer).brand,
+      getAmountMath,
+    };
+    harden(zcf);
+
+    // To Zoe, we will return the invite and an object such that Zoe
+    // can tell us about new seats.
+    /** @type AddSeatObj */
+    const addSeatObj = {
```

Comment:
Why not
```js
const addSeatObj = harden({
```
and kill line 174? (harden typing again?)

See also line 50. Perhaps many other places?